### ISSUE TITLE
Review Ctrl-G HMM+DFA decoder scoring against expected conditional control objective

### ISSUE THREAD

**Issue by SueMarsR** [HLR/DomiKnowS#460]
## Context

The current HMM+DFA generation implementation is in:

- `domiknows/generation/applications/hybrid.py`
- related decoder code under `domiknows/generation/`

This came up while reviewing the Ctrl-G-style implementation. Recent updates mention a lazy/static HMM+DFA success table, so the concern is not only about runtime memoization.

## Observed behavior to review

The current decoder appears to combine several scoring terms by default:

- backend/LM token logits, when enabled
- HMM next-label logits
- HMM+DFA lookahead/success logits

In particular, the default behavior appears to include a standalone HMM next-label contribution in addition to the HMM+DFA lookahead/success contribution.

This should be reviewed against the intended Ctrl-G conditional control objective. The expected Ctrl-G-style behavior conditions the base language model on future constraint satisfaction through a control/lookahead factor. A standalone HMM next-label term may change the effective decoding objective into a different product-style combination.

## Why this matters

If the standalone HMM next-label term is active by default, the decoder may prefer tokens that are likely under the HMM even when all candidate tokens have equal constraint-success probability. This can change decoding behavior independently of DFA constraint satisfaction.

A minimal sanity case to check is:

- two candidate labels/tokens
- equal DFA/HMM lookahead success probability for both candidates
- LM prefers token A
- HMM next-label probability strongly prefers token B

Under a pure conditional control interpretation, the decision should be driven by the LM when the control-success factor is equal. Under the current default combination, the HMM next-label term may flip the decision.

## Recent meeting/update context

The latest meeting update mentions that a lazy static HMM+DFA success table was added and significantly reduced the number of computed HMM states. It also reports an inference run with:

- `hmm_weight=1.0`
- `backend_generator_weight=0.0`
- `lookahead_weight=1.0`
- `dfa_valid=1.0000`
- `exact_sequence=0.0000`
- `token_accuracy=0.8291`

This suggests caching/performance has already been partly addressed, but it does not by itself confirm that the probabilistic decoding objective matches the intended Ctrl-G behavior.

## Requested review

Please verify whether the default HMM+DFA decoder scoring is intended to match standard Ctrl-G-style conditional decoding, or whether it is intentionally implementing a different HMM+LM product-style decoder.

No implementation change is proposed in this issue; this is first to clarify and document the intended objective.

**Comment by auszok** [comment]
Thanks for raising this. I reviewed the HMM+DFA scoring path and the concern was valid.

The previous default `product_hmm_dfa` objective was not a pure Ctrl-G-style conditional control objective. It used a log-linear/product-style blend:

```text
hf_weight * HF_label_logit
+ hmm_weight * log P_HMM(label | h_t)
+ lookahead_weight * log P_HMM(DFA success | h_{t+1}, q_{t+1})
```
I updated the decoder so product_hmm_dfa now defaults to a Ctrl-G-compatible objective:
```text
base_weight * base_model_label_score
+ lookahead_weight * log P_HMM(DFA success | h_{t+1}, q_{t+1})
```

The old product-style objective remains available under the more descriptive name:

```
decode_strategy="hmm_dfa_log_linear"
```

I also added the requested sanity regression:

- equal lookahead success for A and B
- backend/LM prefers A
- HMM next-label strongly prefers B
- default product_hmm_dfa chooses A
- hmm_dfa_log_linear can choose B
